[BUG] TensorUnique operand failed on ray shuffle with AssertionError

<!--
Thank you for your contribution!

Please review https://github.com/mars-project/mars/blob/master/CONTRIBUTING.rst before opening an issue.
-->

**Describe the bug**
Mars unique operand failed on ray shuffle with AssertionError on check `assert len(subtask_chunk_graph.result_chunks) == 1`:
```python
def _get_subtask_out_info(
    subtask_chunk_graph: ChunkGraph, is_mapper: bool, n_reducers: int = None
):
    # output_keys might be duplicate in chunk graph, use dict to deduplicate.
    # output_keys order should be consistent with remote `execute_subtask`,
    # dict can preserve insert order.
    output_keys = {}
    for chunk in subtask_chunk_graph.result_chunks:
        if isinstance(
            chunk.op, VirtualOperand
        ):  # FIXME(chaokunyang) no need to check this?
            continue
        elif is_mapper:
            assert (
                len(subtask_chunk_graph.result_chunks) == 1
            ), subtask_chunk_graph.result_chunks
            assert n_reducers is not None
            return set(), n_reducers
        else:
            output_keys[chunk.key] = 1
    return output_keys.keys(), len(output_keys)
```
For unique operands, reducer may be a mapper too and has outputs which isn' mapper blocks. Ray  task-absed shuffle should support this pattern.
![graphviz (4)](https://user-images.githubusercontent.com/12445254/184132642-a19259fd-43d6-4a27-a033-4aaa97d7586e.svg)


**To Reproduce**
To help us reproducing this bug, please provide information below:
1. Your Python version: 3.8
2. The version of Mars you use: master
3. Versions of crucial packages, such as numpy, scipy and pandas
4. Full stack of the error.
```
FAILED mars/tensor/base/tests/test_base_execution.py::test_unique_execution - AssertionError: [Chunk <op=TensorUnique, stage=reduce, key=f0628e398249f4071de49f3bc018f471_0>, Chunk <o...
============================================================================= 1 failed, 1 warning in 20.37s =============================================================================
2022-08-11 20:01:26,022 ERROR (unknown file):0 -- Task exception was never retrieved
future: <Task finished name='Task-336' coro=<_wrap_awaitable() done, defined at /Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/asyncio/tasks.py:688> exception=AssertionError([Chunk <op=TensorUnique, stage=reduce, key=f0628e398249f4071de49f3bc018f471_0>, Chunk <op=TensorUnique, stage=reduce, key=f0628e398249f4071de49f3bc018f471_1>])>
Traceback (most recent call last):
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/runner.py", line 338, in from_call
    result: Optional[TResult] = func()
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/runner.py", line 259, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
    raise e
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/runner.py", line 166, in pytest_runtest_call
    item.runtest()
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/python.py", line 1761, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/site-packages/_pytest/python.py", line 192, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/tensor/base/tests/test_base_execution.py", line 1013, in test_unique_execution
    res = fetch(*execute(y, inverse))
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/deploy/oscar/session.py", line 1890, in execute
    return session.execute(
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/deploy/oscar/session.py", line 1684, in execute
    execution_info: ExecutionInfo = fut.result(
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/deploy/oscar/session.py", line 1870, in _execute
    await execution_info
  File "/Users/chaokunyang/anaconda3/envs/mars3.8/lib/python3.8/asyncio/tasks.py", line 695, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/deploy/oscar/session.py", line 105, in wait
    return await self._aio_task
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/deploy/oscar/session.py", line 953, in _run_in_background
    raise task_result.error.with_traceback(task_result.traceback)
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/services/task/supervisor/processor.py", line 369, in run
    await self._process_stage_chunk_graph(*stage_args)
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/services/task/supervisor/processor.py", line 247, in _process_stage_chunk_graph
    chunk_to_result = await self._executor.execute_subtask_graph(
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/services/task/execution/ray/executor.py", line 505, in execute_subtask_graph
    output_keys, out_count = _get_subtask_out_info(
  File "/Users/chaokunyang/Desktop/chaokun/python/mars/mars/services/task/execution/ray/executor.py", line 258, in _get_subtask_out_info
    assert (
AssertionError: [Chunk <op=TensorUnique, stage=reduce, key=f0628e398249f4071de49f3bc018f471_0>, Chunk <op=TensorUnique, stage=reduce, key=f0628e398249f4071de49f3bc018f471_1>]
```
5. Minimized code to reproduce the error.
```
 MARS_CI_BACKEND=ray pytest -v -s mars/tensor/base/tests/test_base_execution.py::test_unique_execution
```
```python
@pytest.mark.ray_dag
def test_unique_execution(setup):
    rs = np.random.RandomState(0)
    raw = rs.randint(10, size=(10,))
    x = tensor(raw, chunk_size=3)
    y, inverse = unique(x, return_inverse=True)
    res = fetch(*execute(y, inverse))
```
**Expected behavior**
A clear and concise description of what you expected to happen.

**Additional context**
Add any other context about the problem here.

`DataFrameIndexAlign` operand has similar issue:

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mars/deploy/oscar/session.py:1890: in execute
    return session.execute(
mars/deploy/oscar/session.py:1684: in execute
    execution_info: ExecutionInfo = fut.result(
../../../../anaconda3/envs/mars3.8/lib/python3.8/concurrent/futures/_base.py:444: in result
    return self.__get_result()
../../../../anaconda3/envs/mars3.8/lib/python3.8/concurrent/futures/_base.py:389: in __get_result
    raise self._exception
mars/deploy/oscar/session.py:1870: in _execute
    await execution_info
../../../../anaconda3/envs/mars3.8/lib/python3.8/asyncio/tasks.py:695: in _wrap_awaitable
    return (yield from awaitable.__await__())
mars/deploy/oscar/session.py:105: in wait
    return await self._aio_task
mars/deploy/oscar/session.py:953: in _run_in_background
    raise task_result.error.with_traceback(task_result.traceback)
mars/services/task/supervisor/processor.py:369: in run
    await self._process_stage_chunk_graph(*stage_args)
mars/services/task/supervisor/processor.py:247: in _process_stage_chunk_graph
    chunk_to_result = await self._executor.execute_subtask_graph(
mars/services/task/execution/ray/executor.py:507: in execute_subtask_graph
    output_keys, out_count = _get_subtask_out_info(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

subtask_chunk_graph = <mars.core.graph.entity.ChunkGraph object at 0x7fea91c08e00>, is_mapper = True, n_reducers = 3

    def _get_subtask_out_info(
        subtask_chunk_graph: ChunkGraph, is_mapper: bool, n_reducers: int = None
    ):
        # output_keys might be duplicate in chunk graph, use dict to deduplicate.
        # output_keys order should be consistent with remote `execute_subtask`,
        # dict can preserve insert order.
        output_keys = {}
        for chunk in subtask_chunk_graph.result_chunks:
            if isinstance(
                chunk.op, VirtualOperand
            ):  # FIXME(chaokunyang) no need to check this?
                continue
            elif is_mapper:
>               assert (
                    len(subtask_chunk_graph.result_chunks) == 1
                ), subtask_chunk_graph.result_chunks
E               AssertionError: [DataFrameChunkData <op=DataFrameIndexAlign, stage=map, key=c7a56e28260a78c9b936be2c72feb298_0>, DataFrameChunkData <op=DataFrameIndexAlign, stage=map, key=ceab58328a4b66ef381aceeb58b8cf0f_0>, DataFrameChunkData <op=DataFrameIndexAlign, stage=map, key=c738082cdc61d25a1ed37e5fcabe0b3e_0>]

mars/services/task/execution/ray/executor.py:259: AssertionError

```

reproduction:
` MARS_CI_BACKEND=ray pytest -v -s mars/dataframe/indexing/tests/test_indexing_execution.py::test_align_execution`

--- a/LICENSE
+++ b/LICENSE
@@ -127,7 +127,7 @@ Copyright 1999-2021 Alibaba Group Holding Ltd.
       may provide additional or different license terms and conditions
       for use, reproduction, or distribution of Your modifications, or
       for any such Derivative Works as a whole, provided Your use,
-      reproduction, and distribution of the Work otherwise complies with
+      reproduction, and distribution of the Work otherwise compiles with
       the conditions stated in this License.
 
    5. Submission of Contributions. Unless You explicitly state otherwise,

--- a/benchmarks/asv_bench/benchmarks/serialize.py
+++ b/benchmarks/asv_bench/benchmarks/serialize.py
@@ -47,7 +47,7 @@ from mars.services.subtask import Subtask, SubtaskResult, SubtaskStatus
 from mars.services.task import new_task_id
 from mars.utils import tokenize
 
-# do warmup
+# do warm up
 serialize(None)
 
 

--- a/docs/source/locale/zh_CN/LC_MESSAGES/reference/tensor/generated/mars.tensor.array.po
+++ b/docs/source/locale/zh_CN/LC_MESSAGES/reference/tensor/generated/mars.tensor.array.po
@@ -135,7 +135,7 @@ msgstr ""
 #: mars.tensor.array:34 of
 msgid ""
 "Specifies the minimum number of dimensions that the resulting array "
-"should have.  Ones will be pre-pended to the shape as needed to meet this"
+"should have.  Ones will be prepended to the shape as needed to meet this"
 " requirement."
 msgstr ""
 

--- a/docs/source/locale/zh_CN/LC_MESSAGES/user_guide/tensor/generated/mars.tensor.array.po
+++ b/docs/source/locale/zh_CN/LC_MESSAGES/user_guide/tensor/generated/mars.tensor.array.po
@@ -135,7 +135,7 @@ msgstr ""
 #: mars.tensor.array:34 of
 msgid ""
 "Specifies the minimum number of dimensions that the resulting array "
-"should have.  Ones will be pre-pended to the shape as needed to meet this"
+"should have.  Ones will be prepended to the shape as needed to meet this"
 " requirement."
 msgstr ""
 

--- a/mars/core/entity/chunks.py
+++ b/mars/core/entity/chunks.py
@@ -21,6 +21,9 @@ class ChunkData(EntityData):
     __slots__ = ()
 
     is_broadcaster = BoolField("is_broadcaster", default=False)
+    # If the operand is a shuffle mapper, this flag indicates whether the current chunk is mapper chunk when
+    # the operand produce multiple chunks such as TensorUnique.
+    is_mapper = BoolField("is_mapper", default=False)
     # optional fields
     _index = TupleField("index", FieldTypes.uint32)
 

--- a/mars/learn/linear_model/tests/test_base.py
+++ b/mars/learn/linear_model/tests/test_base.py
@@ -68,7 +68,7 @@ def test_linear_regression(setup):
     assert_array_almost_equal(reg.intercept_, model.intercept_)
     assert_array_almost_equal(reg.predict(X), model.predict(X))
 
-    # Extra case #1: singluar matrix, degenerate input
+    # Extra case #1: singular matrix, degenerate input
     error_msg = re.escape("Does not support sigular matrix!")
 
     X = [[1]]
@@ -78,7 +78,7 @@ def test_linear_regression(setup):
     with pytest.raises(NotImplementedError, match=error_msg):
         reg.fit(X, Y)
 
-    # # Extra case #2: algebrically singluar matrix but algorithmically not
+    # # Extra case #2: algebrically singular matrix but algorithmically not
     # # Works locally but not work in github checks
     # # May be because the inverse is super large
     # X = [[1, 1.5], [1.8, 2]]

--- a/mars/services/task/execution/ray/executor.py
+++ b/mars/services/task/execution/ray/executor.py
@@ -196,18 +196,25 @@ def execute_subtask(
     output = {
         key: data for key, data, _ in iter_output_data(subtask_chunk_graph, context)
     }
+    output_values = []
+    mapper_output = {}
     # assert output keys order consistent
     if is_mapper:
-        chunk_keys = set(k[0] for k in output.keys())
-        assert len(set(chunk_keys)) == 1, chunk_keys
+        # mapper may produce outputs which isn't shuffle blocks, such as TensorUnique._execute_agg_reduce.
+        mapper_main_keys = set()
+        for k, v in output.items():
+            if k not in output_meta_keys:
+                mapper_output[k] = v
+                mapper_main_keys.add(k[0])
+        assert len(mapper_main_keys) == 1, mapper_main_keys
         # sorted reducer_index's consistency with reducer_ordinal is checked in
         # `OperandTilesHandler._check_shuffle_reduce_chunks`.
         # So sort keys by reducer_index to ensure mapper outputs consist with reducer_ordinal,
         # then downstream can fetch shuffle blocks by reducer_ordinal.
-        output = dict(sorted(output.items(), key=lambda item: item[0][1]))
-    output_values = []
+        mapper_output = dict(sorted(mapper_output.items(), key=lambda item: item[0][1]))
+        for mapper_key in mapper_output.keys():
+            output.pop(mapper_key)
     if output_meta_keys:
-        assert not is_mapper
         output_meta = {}
         # for non-shuffle subtask, record meta in supervisor.
         for chunk in subtask_chunk_graph.result_chunks:
@@ -222,6 +229,7 @@ def execute_subtask(
         assert len(output_meta_keys) == len(output_meta)
         output_values.append(output_meta)
     output_values.extend(output.values())
+    output_values.extend(mapper_output.values())
 
     if not is_mapper:
         expect_output_keys, _ = _get_subtask_out_info(subtask_chunk_graph, False)
@@ -250,20 +258,30 @@ def _get_subtask_out_info(
     # output_keys order should be consistent with remote `execute_subtask`,
     # dict can preserve insert order.
     output_keys = {}
+    shuffle_chunk = None
     for chunk in subtask_chunk_graph.result_chunks:
         if isinstance(
             chunk.op, VirtualOperand
         ):  # FIXME(chaokunyang) no need to check this?
             continue
         elif is_mapper:
-            assert (
-                len(subtask_chunk_graph.result_chunks) == 1
-            ), subtask_chunk_graph.result_chunks
-            assert n_reducers is not None
-            return set(), n_reducers
+            if len(subtask_chunk_graph.result_chunks) == 1:
+                return set(), n_reducers
+            elif not chunk.is_mapper:
+                output_keys[chunk.key] = 1
+            else:
+                assert shuffle_chunk is None, (shuffle_chunk, chunk)
+                shuffle_chunk = chunk
         else:
             output_keys[chunk.key] = 1
-    return output_keys.keys(), len(output_keys)
+    output_count = len(output_keys)
+    if is_mapper:
+        assert n_reducers is not None
+        # mapper may produce outputs which isn't shuffle blocks, such as TensorUnique._execute_agg_reduce which is
+        # mapper too, but some outputs are not mapper blocks:
+        # https://user-images.githubusercontent.com/12445254/184132642-a19259fd-43d6-4a27-a033-4aaa97d7586e.svg
+        output_count += n_reducers
+    return output_keys.keys(), output_count
 
 
 @register_executor_cls
@@ -534,15 +552,16 @@ class RayTaskExecutor(TaskExecutor):
                 output_object_refs[0]
             ] = subtask
             if subtask_output_meta_keys:
-                assert not is_mapper
                 meta_object_ref, *output_object_refs = output_object_refs
                 # TODO(fyrestone): Fetch(not get) meta object here.
                 output_meta_object_refs.append(meta_object_ref)
             if is_mapper:
-                shuffle_manager.add_mapper_output_refs(subtask, output_object_refs)
-            else:
-                subtask_outputs = zip(output_keys, output_object_refs)
-                task_context.update(subtask_outputs)
+                shuffle_manager.add_mapper_output_refs(
+                    subtask, output_object_refs[len(subtask_output_meta_keys) :]
+                )
+                output_object_refs = output_object_refs[: len(subtask_output_meta_keys)]
+            subtask_outputs = zip(output_keys, output_object_refs)
+            task_context.update(subtask_outputs)
         logger.info("Submitted %s subtasks of stage %s.", len(subtask_graph), stage_id)
 
         key_to_meta = {}

--- a/mars/services/task/execution/ray/executor.py
+++ b/mars/services/task/execution/ray/executor.py
@@ -227,11 +227,6 @@ def execute_subtask(
         output_values.append(output_meta)
     output_values.extend(normal_output.values())
     output_values.extend(mapper_output.values())
-
-    if not is_mapper:
-        expect_output_keys, _ = _get_subtask_out_info(subtask_chunk_graph, False)
-        output_keys = normal_output.keys()
-        assert expect_output_keys == output_keys, (expect_output_keys, output_keys)
     logger.info("Finish executing subtask %s.", subtask_id)
     return output_values[0] if len(output_values) == 1 else output_values
 
@@ -553,9 +548,9 @@ class RayTaskExecutor(TaskExecutor):
                 output_meta_object_refs.append(meta_object_ref)
             if is_mapper:
                 shuffle_manager.add_mapper_output_refs(
-                    subtask, output_object_refs[len(subtask_output_meta_keys) :]
+                    subtask, output_object_refs[-n_reducers:]
                 )
-                output_object_refs = output_object_refs[: len(subtask_output_meta_keys)]
+                output_object_refs = output_object_refs[:-n_reducers]
             subtask_outputs = zip(output_keys, output_object_refs)
             task_context.update(subtask_outputs)
         logger.info("Submitted %s subtasks of stage %s.", len(subtask_graph), stage_id)

--- a/mars/services/task/execution/ray/executor.py
+++ b/mars/services/task/execution/ray/executor.py
@@ -193,27 +193,24 @@ def execute_subtask(
     # For non-mapper subtask, output context is chunk key to results.
     # For mapper subtasks, output context is data key to results.
     # `iter_output_data` must ensure values order since we only return values.
-    output = {
-        key: data for key, data, _ in iter_output_data(subtask_chunk_graph, context)
-    }
-    output_values = []
+    normal_output = {}
     mapper_output = {}
+    for key, data, is_mapper_block in iter_output_data(subtask_chunk_graph, context):
+        if is_mapper_block:
+            mapper_output[key] = data
+        else:
+            normal_output[key] = data
+    output_values = []
     # assert output keys order consistent
     if is_mapper:
         # mapper may produce outputs which isn't shuffle blocks, such as TensorUnique._execute_agg_reduce.
-        mapper_main_keys = set()
-        for k, v in output.items():
-            if k not in output_meta_keys:
-                mapper_output[k] = v
-                mapper_main_keys.add(k[0])
+        mapper_main_keys = set(k[0] for k in mapper_output.keys())
         assert len(mapper_main_keys) == 1, mapper_main_keys
         # sorted reducer_index's consistency with reducer_ordinal is checked in
         # `OperandTilesHandler._check_shuffle_reduce_chunks`.
         # So sort keys by reducer_index to ensure mapper outputs consist with reducer_ordinal,
         # then downstream can fetch shuffle blocks by reducer_ordinal.
         mapper_output = dict(sorted(mapper_output.items(), key=lambda item: item[0][1]))
-        for mapper_key in mapper_output.keys():
-            output.pop(mapper_key)
     if output_meta_keys:
         output_meta = {}
         # for non-shuffle subtask, record meta in supervisor.
@@ -228,12 +225,12 @@ def execute_subtask(
                 output_meta[chunk_key] = get_chunk_params(chunk), memory_size
         assert len(output_meta_keys) == len(output_meta)
         output_values.append(output_meta)
-    output_values.extend(output.values())
+    output_values.extend(normal_output.values())
     output_values.extend(mapper_output.values())
 
     if not is_mapper:
         expect_output_keys, _ = _get_subtask_out_info(subtask_chunk_graph, False)
-        output_keys = output.keys()
+        output_keys = normal_output.keys()
         assert expect_output_keys == output_keys, (expect_output_keys, output_keys)
     logger.info("Finish executing subtask %s.", subtask_id)
     return output_values[0] if len(output_values) == 1 else output_values
@@ -259,29 +256,28 @@ def _get_subtask_out_info(
     # dict can preserve insert order.
     output_keys = {}
     shuffle_chunk = None
+    if is_mapper:
+        assert n_reducers is not None
+        if len(subtask_chunk_graph.result_chunks) == 1:
+            return set(), n_reducers
+        for chunk in subtask_chunk_graph.result_chunks:
+            if not chunk.is_mapper:
+                output_keys[chunk.key] = 1
+                # mapper may produce outputs which isn't shuffle blocks, such as TensorUnique._execute_agg_reduce
+                # which is  mapper too, but some outputs are not mapper blocks:
+                # https://user-images.githubusercontent.com/12445254/184132642-a19259fd-43d6-4a27-a033-4aaa97d7586e.svg
+            else:
+                assert shuffle_chunk is None, (shuffle_chunk, chunk)
+                shuffle_chunk = chunk
+        return output_keys.keys(), len(output_keys) + n_reducers
     for chunk in subtask_chunk_graph.result_chunks:
         if isinstance(
             chunk.op, VirtualOperand
         ):  # FIXME(chaokunyang) no need to check this?
             continue
-        elif is_mapper:
-            if len(subtask_chunk_graph.result_chunks) == 1:
-                return set(), n_reducers
-            elif not chunk.is_mapper:
-                output_keys[chunk.key] = 1
-            else:
-                assert shuffle_chunk is None, (shuffle_chunk, chunk)
-                shuffle_chunk = chunk
         else:
             output_keys[chunk.key] = 1
-    output_count = len(output_keys)
-    if is_mapper:
-        assert n_reducers is not None
-        # mapper may produce outputs which isn't shuffle blocks, such as TensorUnique._execute_agg_reduce which is
-        # mapper too, but some outputs are not mapper blocks:
-        # https://user-images.githubusercontent.com/12445254/184132642-a19259fd-43d6-4a27-a033-4aaa97d7586e.svg
-        output_count += n_reducers
-    return output_keys.keys(), output_count
+    return output_keys.keys(), len(output_keys)
 
 
 @register_executor_cls

--- a/mars/tensor/base/tests/test_base_execution.py
+++ b/mars/tensor/base/tests/test_base_execution.py
@@ -985,8 +985,7 @@ def test_searchsorted_execution(setup):
             np.testing.assert_array_equal(res, expected)
 
 
-# TODO https://github.com/mars-project/mars/issues/3221
-# @pytest.mark.ray_dag
+@pytest.mark.ray_dag
 def test_unique_execution(setup):
     rs = np.random.RandomState(0)
     raw = rs.randint(10, size=(10,))

--- a/mars/tensor/base/unique.py
+++ b/mars/tensor/base/unique.py
@@ -242,6 +242,8 @@ class TensorUnique(TensorMapReduceOperand, TensorOperandMixin):
             chunks = reduce_op.new_chunks(
                 [shuffle_chunk], kws=kws, order=op.outputs[0].order
             )
+            if op.return_inverse:
+                chunks[0].is_mapper, chunks[1].is_mapper = False, True
             for j, c in enumerate(chunks):
                 reduce_chunks[j].append(c)
 

--- a/mars/tensor/datasource/array.py
+++ b/mars/tensor/datasource/array.py
@@ -233,7 +233,7 @@ def array(x, dtype=None, copy=True, order="K", ndmin=None, chunk_size=None):
         Notes section. The default order is 'K'.
     ndmin : int, optional
         Specifies the minimum number of dimensions that the resulting
-        array should have.  Ones will be pre-pended to the shape as
+        array should have.  Ones will be prepended to the shape as
         needed to meet this requirement.
     chunk_size: int, tuple, optional
         Specifies chunk size for each dimension.

